### DISCUSSION TITLE
Can Kernel environment be ignored?

### DISCUSSION THREAD

**Discussion by freiondrej-lmc** [symfony/symfony#46749] [Q&A] [answered]
Hello,

I've been going through https://12factor.net/config and came to think: would it be possible for a symfony app to be configured without specifying an environment ('dev', 'test', 'prod', ...)? And use specific environment variables to enable / disable different features (often bound to an environment)? E.g. MONOLOG_USE_DEBUG to enable features like https://github.com/symfony/demo/blob/main/config/packages/monolog.yaml#L1-L20, PROFILER_SHOW_TOOLBAR to achieve something like https://github.com/symfony/demo/blob/main/config/packages/web_profiler.yaml#L9-L15 etc.
To achieve this without having to change `symfony/http-kernel` code, there could be `if` statements in `Kernel::configureContainer()` to conditionally add e.g. `config/conditional/profiler_enabled.yml` etc:
```php
if ($_ENV['PROFILER_SHOW_TOOLBAR']) {
   $loader->load($this->getProjectDir() . '/config/conditional/profiler_enabled.yml');
}
```
This approach could also be used in `config/bundles.php`:
```php
$bundles = [
    // some unconditional bundles
    Symfony\Bundle\FrameworkBundle\FrameworkBundle::class => ['all' => true],
    Symfony\Bundle\SecurityBundle\SecurityBundle::class => ['all' => true],
    // ...
];

if ($_ENV['ENABLE_DEBUG_BUNDLE']) {
    $bundles[] = Symfony\Bundle\DebugBundle\DebugBundle::class;
    // or maybe this would be necessary:
    $bundles[Symfony\Bundle\DebugBundle\DebugBundle::class] = ['all' => true];
}

return $bundles;
```
Thanks to this, the knowledge of how the production / development / testing environment (=deployment location) behaves would not be contained in the app, but defined outside of it (e.g. docker etc).
Quoting from the 12factor app page:
> In a twelve-factor app, env vars are granular controls, each fully orthogonal to other env vars. They are never grouped together as “environments”, but instead are independently managed for each deploy. This is a model that scales up smoothly as the app naturally expands into more deploys over its lifetime.

My motivation is to decrease growing complexity of a big project which still uses app.php and app_dev.php files (from SF 3 I believe), which gives some granularity benefits, but at the same time is difficult to manage. Simply migrating to a single app.php and env variables like APP_ENV would probably be too constraining in some aspects.

My questions are:
1) do you think this is a good idea?
2) is there some `Kernel::getEnvironment()`-dependent behavior directly in symfony internals, which would make my approach impossible?

Thank you! 🙂

**Comment by stof** [comment, marked as answer]
using purely environment variable has drawbacks: many configuration setting require a cache clearing to have effect, and so are not supported through Symfony's `%env()%` which is meant to avoid caching the result of reading the env variable. It would be confusing (and not compatible with the 12factor philosophy either) if changing an environment variable would have no effect until you rebuild your app with the new config.

>  production / development / testing environment (=deployment location) 

The Symfony environment is not the same than the deployment location (which is also why we introduced a separate runtime-env concept when introducing secrets, which defaults to the kernel environment but can be configured separately).
for instance, at Incenteev, we have multiple deployment locations (production, staging, review apps per PR) that are all using the production environment.

The kernel environment is similar to the `mode` of webpack for instance, having configuration optimized for production or for dev.

**Reply by freiondrej-lmc** [reply]
@stof I did not think about the caching implications, that makes sense. 🤔 
EDIT: as this answers my original question, I'm marking this as an answer, thank you. I also created a followup question: https://github.com/symfony/symfony/discussions/46757 .
